File: iOS6bookExamples/ch36p906XML/p741p758XML/Person.m

#import "Person.h"


@implementation Person


- (void)encodeWithCoder:(NSCoder *)encoder {
    //[super encodeWithCoder: encoder];
    [encoder encodeObject:self.lastName forKey:@"last"];
    [encoder encodeObject:self.firstName forKey:@"first"];
}

- (id) initWithCoder:(NSCoder *)decoder {
    //self = [super initWithCoder: decoder];
    self = [super init];
    self->_lastName = [decoder decodeObjectForKey:@"last"];
    self->_firstName = [decoder decodeObjectForKey:@"first"];
    return self;
}

- (NSString*) description {
    return [NSString stringWithFormat: @"%@ %@", self.firstName, self.lastName];
}

@end

File: iOS13bookExamples/bk2ch23p798basicFileOperations/ch36p1053basicFileOperations/Person.swift


import UIKit

// adopt secure coding for iOS 12

class Person: NSObject, NSSecureCoding {
    static var supportsSecureCoding: Bool { return true }
    
    var firstName : String
    var lastName : String
    
    override var description : String {
        return self.firstName + " " + self.lastName
    }
    
    init(firstName:String, lastName:String) {
        self.firstName = firstName
        self.lastName = lastName
        super.init()
    }
    
    func encode(with coder: NSCoder) {
        // do not call super in this case
        coder.encode(self.lastName, forKey: "last")
        coder.encode(self.firstName, forKey: "first")
    }
    
    required init(coder: NSCoder) {
        self.lastName = coder.decodeObject(of: NSString.self, forKey:"last")! as String
        self.firstName = coder.decodeObject(of: NSString.self, forKey:"first")! as String
        // do not call super init(coder:) in this case
        super.init()
    }
}

// but in Swift 4 we are more likely to do it this way

class Person2: NSObject, Codable {
    var firstName : String
    var lastName : String
    
    override var description : String {
        return self.firstName + " " + self.lastName
    }
    
    init(firstName:String, lastName:String) {
        self.firstName = firstName
        self.lastName = lastName
        super.init()
    }
}

